rdd/node-microservices.md

```diff
@@ -0,0 +1,98 @@
+
+## Plan
+
+1. Add `/src/${moduleName}/index.ts` to each module
+1. Update package.json
+1. Update docker-compose.yml
+1. Helpers
+
+### 1. Add `index.ts` to each module
+
+Each module's `index.ts` will be similar to our current `index.ts`, except that it'll only instantiate that particular module instead of all of them.
+
+This file will also be responsible for loading configuration, unlike our current approach in which `app` takes care of this. 
+
+As the current `index.ts`, each `index.ts` will need to import side-effecting files (those under `Extensions`, for example).
+
+### 2. Update `package.json`
+
+The Node's `package.json` currently has a `bin` section that maps `poet-node` to `dist/babel/src/index.js`.
+
+This is used by NPM at installation time, when running `npm i -g @po.et/node` or `cd node & npm i -g`, to set up a symbolic link that allows running `poet-node` in the shell, the same way we'd run `cat`, `vim`, `bitcoind` or any other program.
+
+We'll want to add several entries to the `bin` section, one for each module, mapping module names to the newly created `module/index.ts` files. For example: `"poet-node-blockchain-writer": "dist/babel/src/BlockchainWriter/index.js"`
+
+### 3. Update `docker-compose.yml`
+
+We will need to remove the current `poet-node` entry from `docker-compose.yml` and in its place add one entry for each module.
+
+The `command` of the docker-compose service for each module would be almost the same as the one we have for `poet-node`, the only difference being `npm start` being replaced by `poet-node-{module}`.
+
+All docker-compose module services will need to be linked to `rabbit`, but not to other module services.
+
+The `dockerfile` will need to `npm i -g` to make the `poet-node-{module}` commands available..
+
+### 4. Helpers
+
+Everything under `Helpers` will go into its own library, `@po.et/helpers`, for two reasons:
+- they would not be directly available to the modules any more if we actually separated them into their own projects with their own `package.json`'s, 
+- most (if not all) of the functions there are not specific to the Po.et Node and many have already been copy-pasted to Frost API.
+
+Other helper functions that may exist in other projects but not in the Node could also be migrated to this new package.
+
+## Tests
+
+In this first step we will stop using `/src/index.ts` in production, but keep `/src/app.ts` for tests. 
+
+In the future we will need to re-think how we approach our tests. 
+
+### Unit Tests
+
+These should require no change at all.
+
+### Integration Tests
+
+Since each module will actually be a separate application, integration tests will be aimed at specific modules and live inside their code base. 
+
+The preferred approach towards this model is to _mock_ all other modules, triggering specific behaviors by feeding the module being tested prepared RabbitMQ messages and listening to the messages it publishes.
```

Comment:
Testing via RabbitMQ messages couples us to RabbitMQ. I think testing from the boundary at the controller using inversion of control would be more flexible and easier to test. That means we will need a RouterIn/RouterOut instead of a single Router. The functions in RouterIn/Out can make use of pure functions that can be unit tested to map between RabbitMQ messages and our module messages, so we shouldn't really need integration tests around this area. I can expand more on this later with a diagram, example refactor, etc.

Reply:
That would be two different kind of tests. I think both would be valid. But yeah, let's discuss on the side.

Reply:
How are you defining integration tests? Communication between two to three modules without describing an entire workflow is how I see them. That could be BlockchainWriter writing to mongo and sending a message to rabbit. Have we a good demarcation here?

Reply:
It sounds like we're defining integration tests as from point of entry into a module to point of exit.

Reply:
I used the term _integration tests_ vaguely here. Kenny's suggestion [here](https://github.com/poetapp/documentation/issues/26) is spot on. 

Thinking of it, _functional tests_ is more accurate if following my definition [here](https://github.com/poetapp/documentation/issues/26#issuecomment-443349674).

I'll update the RDD though, replacing "integration" with "functional" for now, and adding a note to make it clear that other kinds of tests or approaches to testing are not discouraged and that terminology isn't well defined yet.

Reply:
Updated